src/chainstate/stacks/boot/bns.clar

```diff
@@ -0,0 +1,973 @@
+    (asserts! (is-eq (len hashed-salted-fqn) u20) (err ERR_NAME_HASH_MALFORMED))
+    ;; Ensure that user will be burning a positive amount of tokens
+    (asserts! (> stx-to-burn u0) (err ERR_NAME_STX_BURNT_INSUFFICIENT))
+    ;; Burn the tokens
+    (unwrap! (stx-transfer? stx-to-burn contract-caller BURN_ADDRESS) (err ERR_INSUFFICIENT_FUNDS))
+    ;; Register the pre-order
+    (map-set name-preorders
+      ((hashed-salted-fqn hashed-salted-fqn) (buyer contract-caller))
+      ((created-at block-height) (stx-burned stx-to-burn) (claimed false)))
+    (ok (+ block-height NAME_PREORDER_CLAIMABILITY_TTL))))
+
+;; NAME_REGISTRATION
+;; This is the second transaction to be sent. It reveals the salt and the name to all BNS nodes,
+;; and assigns the name an initial public key hash and zone file hash
+(define-public (name-register (namespace (buff 19))
+                              (name (buff 16))
+                              (salt (buff 20))
+                              (zonefile-hash (buff 20)))
+  (let (
+    (hashed-salted-fqn (hash160 (concat (concat (concat name 0x2e) namespace) salt)))
+    (name-currently-owned (map-get? owner-name ((owner contract-caller)))))
+    (let ( 
+        (preorder (unwrap!
+          (map-get? name-preorders ((hashed-salted-fqn hashed-salted-fqn) (buyer contract-caller)))
+          (err ERR_NAME_PREORDER_NOT_FOUND)))
+        (namespace-props (unwrap!
+          (map-get? namespaces ((namespace namespace)))
+          (err ERR_NAMESPACE_NOT_FOUND)))
+        (current-owner (nft-get-owner? names (tuple (name name) (namespace namespace))))
+        (can-sender-register-name (if (is-none name-currently-owned)
+                                 true
+                                  (unwrap! 
+                                    (is-name-lease-expired?
+                                      (unwrap! (get namespace name-currently-owned) (err ERR_PANIC))
+                                      (unwrap! (get name name-currently-owned) (err ERR_PANIC)))
+                                    (err ERR_PANIC)))))
+      ;; The name must only have valid chars
+      (asserts!
+        (not (has-invalid-chars name))
+        (err ERR_NAME_CHARSET_INVALID))
+      ;; The name must not exist yet, or be expired
+      (if (is-none current-owner)
+        true
+        (asserts!
+          (unwrap! (is-name-lease-expired? namespace name) (err ERR_PANIC))
+          (err ERR_NAME_UNAVAILABLE)))
+      ;; The name's namespace must be launched
+      (asserts!
+        (is-some (get launched-at namespace-props))
+        (err ERR_NAMESPACE_NOT_LAUNCHED))
+      ;; The preorder must have been created after the launch of the namespace
+      (asserts!
+        (> (get created-at preorder) (unwrap! (get launched-at namespace-props) (err ERR_PANIC)))
+        (err ERR_NAME_PREORDERED_BEFORE_NAMESPACE_LAUNCH))
+      ;; The preorder entry must be unclaimed - todo(ludo): is this assertion redundant?
```

Comment:
I think this assertion is _necessary_.  Otherwise, in a namespace with a short name lifetime (i.e. less than `NAME_PREORDER_CLAIMABILITY_TTL` blocks), a principal could re-use the same preorder to register the name over and over again.  The preorder should only be used by at most one registration.

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/2013/commits/c04a950cb317589b3234f540b4aba26d1d814067